src/roman_datamodels/stnode.pyi

```diff
@@ -0,0 +1,543 @@
+        reftype: enums.LinearityRefType
+        input_units: base.Unit[u.DN]
+        output_units: base.Unit[u.DN]
+
+    class LinearityRef(stnode.TaggedObjectNode):
+        meta: _LinearityRefMeta
+        coeffs: np.ndarray[3, np.float32]
+        dq: np.ndarray[2, np.uint32]
+
+    class _MaskRefMeta(_RefCommon):
+        reftype: enums.MaskRefType
+
+    class MaskRef(stnode.TaggedObjectNode):
+        meta: _MaskRefMeta
+        dq: np.ndarray[2, np.uint32]
+
+    class _PixelareaRefPhotometry(stnode.DNode):
+        pixelarea_steradians: u.Quantity[u.sr] | None
+        pixelarea_arcsecsq: u.Quantity[u.arcsec**2] | None
+
+    class _PixelareaRefMeta(_RefCommon):
+        reftype: enums.PixelareaRefType
+        photometry: _PixelareaRefPhotometry
+        instrument: _Instrument
+
+    class PixelareaRef(stnode.TaggedObjectNode):
+        meta: _PixelareaRefMeta
+        data: np.ndarray[2, np.float32]
+
+    class _ReadnoiseRefMeta(_RefCommon):
+        reftype: enums.ReadnoiseRefType
+        exposure: _RefExposureType
+
+    class ReadnoiseRef(stnode.TaggedObjectNode):
+        meta: _ReadnoiseRefMeta
+        data: base.Quantity[np.ndarray[2, np.float32], u.DN]
+
+    class _SaturationRefMeta(_RefCommon):
+        reftype: enums.SaturationRefType
+
+    class SaturationRef(stnode.TaggedObjectNode):
+        meta: _SaturationRefMeta
+        data: base.Quantity[np.ndarray[2, np.float32], u.DN]
+        dq: np.ndarray[2, np.uint32]
+
+    class _SuperbiasRefMeta(_RefCommon):
+        reftype: enums.SuperbiasRefType
+
+    class SuperbiasRef(stnode.TaggedObjectNode):
+        meta: _SuperbiasRefMeta
+        data: np.ndarray[2, np.float32]
+        dq: np.ndarray[2, np.uint32]
+        err: np.ndarray[2, np.float32]
+
+    class _WfiImgPhotomRefMeta(_RefCommon):
+        reftype: enums.WfiPhotomRefType
+
+    class _PhotTableEntry(stnode.DNode):
+        photmjsr: u.Quantity[u.MJy / u.sr] | None
+        uncertainty: u.Quantity[u.MJy / u.sr] | None
+        pixelareasr: u.Quantity[u.sr] | None
+
+    class _PhotTable(stnode.DNode):
+        F062: _PhotTableEntry
+        F087: _PhotTableEntry
+        F106: _PhotTableEntry
+        F129: _PhotTableEntry
+        F146: _PhotTableEntry
+        F158: _PhotTableEntry
+        F184: _PhotTableEntry
+        F213: _PhotTableEntry
+        GRISM: _PhotTableEntry
+        PRISM: _PhotTableEntry
+        DARK: _PhotTableEntry
+
+    class WfiImgPhotomRef(stnode.TaggedObjectNode):
+        meta: _WfiImgPhotomRefMeta
+        phot_table: _PhotTable
+
+    class _AssociationProduct(stnode.DNode):
+        name: str
+        members: stnode.LNode[str]
+
+    class Associations(stnode.TaggedObjectNode):
+        asn_id: str
+        asn_pool: str
+        asn_type: str
+        asn_rule: str
+        version_id: str
+        code_version: str
+        degraded_status: str
+        program: int
+        target: int
+        constraints: str
+        products: stnode.LNode[_AssociationProduct]
```

Comment:
I presume this would need upkeep as well?

Reply:
We already essentially do the same work with `factories.py` and `maker_utils.py`, I specifically built the constructor here to be able to fill it out with random data (like `factories.py`) or fixed data (`maker_utils.py`). This way I could replace both modules with construction from a singular, verified source. The duplication of effort between those two modules should have never occurred, instead both modes should have been consolidated.

This presumes of course, that we drop Python 3.9 support. Which I would argue is not a bad idea. According to [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) support for Python 3.9 will be dropped on April 5, 2024. We will have to drop Python 3.9 support at that point because both NumPy and astropy will drop support for it then. This is well before any timeline which would actually operationally use the Roman pipeline in any meaningful fashion. Indeed, I would argue that we only support Python 3.11 since it is the only currently available version of Python which we will be still supported by NumPy/astropy by 2026 (Its support will be dropped April 24, 2026).

Reply:
Note that any replacement for `factories.py` and/or `maker_utils.py` by such a centralized constructor would simply map the interfaces in those two modules back to the interfaces provided by the centralized constructor. So for all intents and purposes, this type of change would not effect any outside code due to interface changes.

Reply:
Okay - I was just asking to gauge what future support would entail.